Turn off locale handling if conditions not met

khwilliamson opened the pull request:
This #if ... #elif ... sequence ended in an #elif, neglecting the
possibility that none of the conditions are true.  This commit adds a
final #else to account for that eventuality.

And what that means should that final #else be taken is that the
platform under the given Configuration can't handle locales; so they are
turned off.

This happens on z/OS under threading.  The OS refuses to change the
locale at all once a second thread is started.  Perl relies on the
ability to change the locale at will to accommodate locales that have a
comma be the radix (decimal point) character and the huge amount of code
that has been written expecting that at all times a dot is the radix
character.  This includes core code (which we could change) and XS code
(which would be overwhelming to change).

Thus Perl won't work on many locales on z/OS when two or more threads
are running.  The only reasonable option is to just turn off the whole
thing.

* This set of changes requires a perldelta entry, to be written

jkeenan commented on perl.h:
```diff
@@ -1175,9 +1166,20 @@ violations are fatal.
+#include "locale_table.h"
+
+#include "perl_langinfo.h"    /* Needed for _NL_LOCALE_NAME */
+
```
You are moving 9 lines of code here to be inside the section said to be `unfortunately duplicated in makedef.pl.`  However, I cant' find any of the following strings anywhere in `makedef.pl`.
```
$ grep -n -E '(PERL_LOCALE_TABLE_ENTRY|locale_table|perl_langinfo)' makedef.pl
$ 
```
Can you clarify?

tonycoz replied:
This chunk should probably be after the "end of makedef.pl logic duplication" immediately below this.

bulk88 commented on makedef.pl:
```diff
@@ -145,6 +145,9 @@ BEGIN
         $define{USE_LOCALE} = 1;
     }
+    else {
+        $define{NO_LOCALE} = 1;
```
What is with the `USE_LOCALE` vs `NO_LOCALE` identifiers in this `.pl` or in the whole build process? What if both are defined by the end user on the command line, or 1 automatically, and 1 by the user?

`makedef.pl` has 2 existing conditionals for `USE_LOCALE`, and none for `NO_LOCALE`.  `makedef.pl` ignores whatever `NO_LOCALE` means, and only cares about `&& $define{USE_LOCALE}` test.

Why suddenly make `NO_LOCALE` defined and `== 1` instead of force undef-ing `USE_LOCALE` or `croak()`ing?

khwilliamson committed:
regen/locale.pl: Use 'USE_LOCALE" not 'NO_LOCALE'

NO_LOCALE is a Configure option used to override setting the internal
option USE_LOCALE.  Internal checks should use the internal version.

Not doing this caused z/OS to think it should be using locales on
threaded systems, whereas it shouldn't.

FYI, the reason this is is that the OS refuses to change the locale at
all once a second thread is started.  Perl relies on the ability to
change the locale at will to accomodate locales that have a comma be the
radix (decimal point) character and the huge amount of code that has
been written expecting that at all times a dot is the radix character.
This includes core code (which we could change) and XS code (which would
be ovewhelming to change).

khwilliamson committed:
perl.h: Make sure USE_LOCALE is off if no locales available

It shouldn't be getting set anyway, but this makes it clear, and also
makes sure someone hasn't tried to define it with Configure
